### ISSUE TITLE
False negative

### ISSUE THREAD

**Issue by ThomasFaivre** [PyCQA/pylint#1707]
Hello!

I believe I have found a small issue with the unsupported-assignment-operation check:

### Steps to reproduce
1. cat test.py 
```
class SmartDict(dict):
    def __getattr__(self, key):
        try:
            return self[key]
        except KeyError:
            raise AttributeError(key)

    def __setattr__(self, key, value):
        self[key] = value

    def __delattr__(self, key):
        try:
            del self[key]
        except KeyError:
            raise AttributeError(key)

a = SmartDict()
a.d = 3
b = SmartDict(d={})
b.d[5] = 42
```
2. pylint --disable=all --enable=unsupported-assignment-operation test.py

### Current behavior
************* Module test
E: 20, 0: 'b.d' does not support item assignment (unsupported-assignment-operation)

### Expected behavior
No error.

### pylint --version output
upstream, sha1: dd78aedcf0f4
pylint 1.8.0, 
astroid 1.5.3
Python 2.7.9 (default, Jun 29 2016, 13:08:31) 
[GCC 4.9.2]

Pylint seems to believe 'd' is an attribute of the SmartDict because of 'a' and does not allow d to be a dict afterward.

**Comment by PCManticore** [comment]
This is somewhat a duplicate of https://github.com/PyCQA/pylint/issues/553, although it's not clear at first. The problem here is two-fold:

- first pylint can't understand your class, since it relies heavily on dynamic features, `__setattr__`, custom `__getattr__` etc.
- but it emits the warning you are referencing because it's instance inference support is somewhat dumb. When it detects a modification for an instance, let's say as `a.d = 3`, it automatically adds that member to all the potential instances of your instance's class, which means that when pylint infers `b.d`, which is a totally separate instance, it will infer as being `3`, which of course does not  support that assignment operation. Dropping the assignment of `d` to `3` should make pylint not complain any longer.

Unfortunately fixing this bug requires a bit of reworking of astroid's internals, which is not quite trivial, which is why that issue is open for a couple of years now.
